mrmustard/lab_dev/circuit_components.py

```diff
@@ -717,21 +754,52 @@ def __matmul__(self, other: CircuitComponent | Scalar) -> CircuitComponent:
             >>> att = Attenuator([0], 0.5)
             >>> assert (coh @ att).wires.input.bra  # the input bra is still uncontracted
         """
+
+
         if isinstance(other, (numbers.Number, np.ndarray)):
             return self * other
 
         wires_result, perm = self.wires @ other.wires
         idx_z, idx_zconj = self._matmul_indices(other)
-        if type(self.representation) == type(other.representation):
+        
+        if type(self.representation) == type(other.representation) == Fock:
             self_rep = self.representation
             other_rep = other.representation
         else:
-            self_rep = self.to_bargmann().representation
-            other_rep = other.to_bargmann().representation
+            index_self, index_other = self.wires.contracted_indices(other.wires)
+            self_rep = self.to_bargmann(index_self).representation
+            other_rep = other.to_bargmann(index_other).representation
 
         rep = self_rep[idx_z] @ other_rep[idx_zconj]
         rep = rep.reorder(perm) if perm else rep
-        return CircuitComponent._from_attributes(rep, wires_result, None)
+        result = CircuitComponent._from_attributes(rep, wires_result, None)
+        result._helper_update_wire_rep(other)
+
+        return result
+
+    def _helper_update_wire_rep(self,other):
+        
+        from .circuit_components_utils import BtoQ, BtoPS
+        if isinstance(other, BtoQ):
+            btoq_param = math.atleast_1d(other.phi.value)
+            if other.wires.bra:
+                for k, m in enumerate(other.modes):
+                    i = self.wires.index_dicts[0][m]
+                    self._index_representation[i] = ('Q', float(btoq_param[k]))
```

Comment:
We spoke about this briefly offline but I think an `Enum` is a much safer way of keeping track of the Q and PS conversions.